Enabled Native Windows Build of System.IO.Compression.Native (zlib)

ianhays opened the pull request:
This commit adds scripts to enable building Windows-specific binaries in CoreFX. The code is mostly modeled off of CoreCLR with some elements taken from build.sh building for native Unix components. It uses CMake to produce the projects that are then passed to msbuild. The logic primarily resides within the "build-native.cmd" script that is called by "build.cmd" from the corefx root when "__BuildNative" is set.

Some things to be aware of:
- __BuildNative in build.cmd is currently always set. Either some logic should be added to actually evaluate it, or the variable should be removed.
- Cross compilation isn't set up. build-native.cmd can only build for windows at this time. It can build for x64 or x86, however. The default is currently x64. AnyCPU also defaults to x64.
- Since manual parsing of args is used, most of the msbuild targets that can be passed to build.cmd will be ignored by native compilation. Currently, only Configuration and Platform are acknowledged. I can imagine a system where the native project can be used with the same targets as the managed ones, but that capability isn't complete now. Those properties/targets are still used in the managed build, of course.
- A TFS-side commit to the package definitions will need to follow this PR so that we can consume the publicly built System.IO.Compression.Native.
- I minimized most of the compilation options for Windows cmake. Some will likely needed to be reenabled in the CMakeLists.txt in corefx/src/native/
- I tested the System.IO.Compression.Native.dll built by "build.cmd /p:Configuration=Release" by renaming it to "clrcompression.dll" and running the unit tests using it.
- This PR will need to be modified pending the merging of https://github.com/dotnet/corefx/pull/4440.
- The Zlib code included is unaltered 1.2.3 code.
- This will unblock https://github.com/dotnet/corefx/issues/3986 @bjjones if we merge this and don't decide to instead consume a different zlib package (e.g. http://nuget.org/api/v2/package/zlib/1.2.8.8)

resolves #826

@stephentoub @mellinoe @joshfree @richlander @eerhardt

stephentoub commented on src/Native/System.IO.Compression.Native/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+project(System.IO.Compression.Native)
```
Currently all of the source under src/Native is Unix-specific.  This PR is adding Windows code as well, but there's nothing to differentiate it.  I suggest before we do this PR that we move the code around to ensure a proper separation, e.g. src/Native/Windows/LibraryName,src/Native/Unix/LibraryName, etc.  Without that, for example, it looks like this PR is going to conflict with @eerhardt's PR that's also adding a System.IO.Compression.Native, except his is for Unix.

ianhays replied:
Yeah, my plan was to modify the CMakeLists to do different things depending on the Target OS - Unix would execute Eric's code, Windows would execute mine. 

The shared code between the two is narrow enough that separate CMakeLists may be better as you said.

stephentoub commented on src/Native/System.IO.Compression.Native/ZLib-1.2.3/ChangeLog:
```diff
@@ -0,0 +1,855 @@
+
+                ChangeLog file for zlib
+
+Changes in 1.2.3 (18 July 2005)
```
I understand this is a meant to be just a step in the journey, and I see you commented in the PR description about possible next steps.  It's still a bit sad that we're still moving forward with an implementation that's over 10 years old when a much newer solution is available.  It'd be great to use newer source at the very least, and ideally just use some official binary distribution instead.  I like the idea of just referencing the latest in the NuGet feed and using that, rather than building a copy from decade-old source.

ianhays replied:
For the record, I agree. I have no plans on leaving it at 1.2.3; I wanted something up sooner rather than later to unblock https://github.com/dotnet/corefx/issues/3986 and get this out in the open. As you said, just a step in the journey :)

> It'd be great to use newer source at the very least

This should be trivial. I'll go ahead and put in an OSS request for 1.2.8 (latest). Replacing 1.2.3 with 1.2.8 will be a copy-paste job and maybe some variable modifications in the CMakeList. I'll know better when approval comes. We can hold off on this until 1.2.8 approval comes, if you'd prefer.

>ideally just use some official binary distribution instead

We can still do this. Adler uploads a windows x86/x64 package to NuGet that works out-of-the-box with System.IO.Compression. This locks us out of making any native modifications, but there's always the argument that we can submit those as a PR to the zlib github if we want them badly enough. Same goes for support for other architectures. Not sure if we could do that with the Intel mods that bjjones was talking about adding, but I haven't looked into that.

ianhays commented:
Sorry for the delay. I've updated the PR with 1.2.8 code. I included all files (half of which we don't use) from the 1.2.8 tar that can mostly be removed once we nail down how we want the native build to work. I've rebased everything to make reviewing easier:

- In the first commit I moved all existing native build code to src/Native/Unix.
- In the second commit I add the zlib code.
- In the third commit I add the Windows native build code.

akoeplinger commented:
Any reason why the version number is included in the folder name? That'll make updating more cumbersome.

ianhays commented:
@akoeplinger The folder is named 1.2.8 because the zip was named 1.2.8. Naming it "zlib" would be better though, thanks. I'll change it in the next feedback update.

ianhays commented:
Native Unix build failed with "src/Native/Unix/gen-buildsys-clang.sh: Permission denied".

ianhays commented:
I've taken some more time to look into the msbuild/vs dependency and consider the options. The documentation for the inner workings of the cmake generators is very slim, but I've got a better idea of how they're used now. There isn't any way to circumvent the necessity for a generator, so we'll need to pick one from our options list. For Windows, we've got:
- Clang
   - The Unix native build relies upon the existence of Clang which is passed as a generator to cmake. I guess we could use Clang for Windows as well, but cursory searching suggests that it isn't fully "there" yet.
- NMake
  - We could produce makefiles for NMake but I believe you need VS to get nmake anyways, so we may as well just use VS then and be consistent with CoreCLR. 
- MinGW
  - There is also a generator for MinGW but that's a completely new dependency to introduce so it's no better than using VS. 
- Ninja
  - Same as above.
- Visual Studio
  - What I used in this PR

I think we should continue using VS for now and stay inline with what CoreCLR does for its native Windows build. It's the most frequently used (by far) generator for Windows and has the best documentation. Changing the generator should be <i>relatively</i> simple in the future when we switch to the core version of msbuild, though I'm not sure how different it will end up being. Is the end goal to be able to build on Windows using msbuild without requiring a VS installation?

ianhays commented:
This PR is in the final steps of the review stage on the TFS side. Closing this PR since I'll be submitting it there and mirroring it over.
